proposals/lip-0041.md

```diff
@@ -97,23 +103,216 @@ In this section, we describe the properties of an auth account and specify their
+
+1. `numberOfSignatures`: The number of private keys that must sign a transaction.
+2. `mandatoryKeys`: An array of public keys. A valid public key is 32 bytes long. Once the account is registered as a multisignature account, every outgoing transaction requires a signature for every public key in `mandatoryKeys`.
+3. `optionalKeys`: An array of public keys. A valid public key is 32 bytes long. Once the account is registered as a multisignature account, every outgoing transaction requires some signatures for some public keys in `optionalKeys` (the number of needed signatures depends on the `numberOfSignatures` property and may also be zero).
+4. `signatures`: An array of signatures, corresponding to the public keys contained in `mandatoryKeys` and `optionalKeys`. All public keys must have a corresponding signature. The signatures corresponding to `mandatoryKeys` are appended first, followed by those corresponding to `optionalKeys`, where the order of signatures is the same as the order of public keys in the respective array. The message that is signed contains the parameters of the transaction, serialized using the `multisigRegMsgSchema` schema given below.
+
+```java
+multisigRegMsgSchema = {
+    "type": "object",
+    "required": ["address", "nonce", "numberOfSignatures", "mandatoryKeys",  "optionalKeys"],
+    "properties": {
+        "address": {
+            "dataType": "bytes",
+            "length" : 20
+            "fieldNumber": 1
+        },
+        "nonce": {
+            "dataType": "uint32",
+            "fieldNumber": 2
+        },
+        "numberOfSignatures": {
+            "dataType": "uint32",
+            "fieldNumber": 3
+        },
+        "mandatoryKeys": {
+            "type": "array",
+            "items": {
+                "dataType": "bytes",
+                "length" : 32
+            },
+            "fieldNumber": 4
+        },
+        "optionalKeys": {
+            "type": "array",
+            "items": {
+                "dataType": "bytes",
+                "length" : 32
+            },
+            "fieldNumber": 5
+        }
+    }
+}
+```
 
 ##### Verification
 
-The verification of this command is defined in [LIP 0017](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0017.md#multisignature-account-registration).
+These specifications supersede those defined in [LIP 0017](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0017.md#multisignature-account-registration).
+
+```python
+def verify(trs: Transaction) -> None:
+    # Check that a multisignature group is not already registered for this address
+    senderAddress = 20-byte address derived from trs.senderPublicKey
+    if len(authAccount(senderAddress).mandatoryKeys) > 0 or len(authAccount(senderAddress).mandatoryKeys) > 0:
+        raise Exception('Multisignature group already registered for this address.')
+    # Mandatory keys are pairwise distinct and ordered lexicographically.
```

Comment:
Maybe we can say that the `mandatoryKeys` array may be empty (this is implied by the Python code, but maybe best o mention it explicitly).

Reply:
added